### PR TITLE
Handle Pipewire Start and Stop Function

### PR DESCRIPTION
**Pull request by dorindabassey**

### CODE REVIEW

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 1]
```diff
@@ -408,4 +408,30 @@ impl AudioBackend for PwBackend {
...
     }
+
+    fn start(&self, stream_id: u32) -> Result<()> {
+        println!("pipewire start");
+        self.thread_loop.lock();
+        let stream_hash = self.stream_hash.read().unwrap();
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.set_active(true).expect("could not start stream");
+        } else {
+            println!("stream not found for the given stream_id");
```
Comment (stefano-garzarella) [inline crates/sound/src/audio_backends/pw_backend.rs]: Should we return an error as result?

File: crates/sound/src/vhu_sound.rs [hunk 2]
```diff
@@ -410,13 +410,58 @@ impl VhostUserSoundThread {
...
+                VIRTIO_SND_R_PCM_START => {
                     let pcm_hdr = desc_chain
                         .memory()
                         .read_obj::<VirtioSoundPcmHeader>(desc_request.addr())
                         .map_err(|_| Error::DescriptorReadFailed)?;
-                    let stream_id: usize = u32::from(pcm_hdr.stream_id) as usize;
-                    dbg!("stream_id: {}", stream_id);
+                    let stream_id = u32::from(pcm_hdr.stream_id);
+
+                    let state = audio_backend.get_stream_state(stream_id).unwrap();
```
Comment (stefano-garzarella) [inline crates/sound/src/vhu_sound.rs]: We forgot to handle this error, but I think we can do a patch on top to fix it.

@MatiasVara WDYT?

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 3]
```diff
@@ -389,4 +389,30 @@ impl AudioBackend for PwBackend {
 
         Ok(())
     }
+
+    fn release(&self, stream_id: u32) -> Result<()> {
+        println!("pipewire backend, release function");
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pw_backend.rs]: should probably use log::debug here.

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 4]
```diff
@@ -389,4 +389,30 @@ impl AudioBackend for PwBackend {
...
+        self.thread_loop.lock();
+        let mut stream_states = self.stream_states.write().unwrap();
+        let stream_hash = self.stream_hash.read().unwrap();
+        let stream_listener = self.stream_listener.read().unwrap();
+
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.disconnect().expect("could not disconnect stream");
+        } else {
+            error!("stream not found for the given stream_id");
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pw_backend.rs]: Shouldn't this also return an error ?

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 5]
```diff
@@ -389,4 +389,30 @@ impl AudioBackend for PwBackend {
...
+        };
+        if let Some(stream_lis) = stream_listener.get(&stream_id) {
+            //unregister stream
+            std::mem::drop(stream_lis);
+        } else {
+            error!("stream listener not found for the given stream_id");
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pw_backend.rs]: same here

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 6]
```diff
@@ -415,4 +415,30 @@ impl AudioBackend for PwBackend {
         self.thread_loop.unlock();
         Ok(())
     }
+
+    fn start(&self, stream_id: u32) -> Result<()> {
+        println!("pipewire start");
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pw_backend.rs]: should be a debug!

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 7]
```diff
@@ -415,4 +415,30 @@ impl AudioBackend for PwBackend {
...
+        self.thread_loop.lock();
+        let stream_hash = self.stream_hash.read().unwrap();
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.set_active(true).expect("could not start stream");
+        } else {
+            error!("stream not found for the given stream_id");
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pw_backend.rs]: same as earlier, should probably return an error

File: crates/sound/src/vhu_sound.rs (hunk 2 above)
Comment (MatiasVara) [reply crates/sound/src/vhu_sound.rs]: I created #21 to not forget to deal correctly with the handling of the errors.

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 8]
```diff
@@ -415,4 +415,30 @@ impl AudioBackend for PwBackend {
...
+
+    fn stop(&self, stream_id: u32) -> Result<()> {
+        println!("pipewire stop");
+        self.thread_loop.lock();
+        let stream_hash = self.stream_hash.read().unwrap();
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.set_active(false).expect("could not stop stream");
+        } else {
+            error!("stream not found for the given stream_id");
+        };
```
Comment (MatiasVara) [inline crates/sound/src/audio_backends/pw_backend.rs]: Why is this method not setting the stream state to VIRTIO_SND_R_PCM_STOP?

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 9]
```diff
@@ -415,4 +415,30 @@ impl AudioBackend for PwBackend {
...
+
+    fn start(&self, stream_id: u32) -> Result<()> {
+        println!("pipewire start");
+        self.thread_loop.lock();
+        let stream_hash = self.stream_hash.read().unwrap();
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.set_active(true).expect("could not start stream");
+        } else {
+            error!("stream not found for the given stream_id");
+        };
```
Comment (MatiasVara) [inline crates/sound/src/audio_backends/pw_backend.rs]: Why is this method not setting the stream state to VIRTIO_SND_R_PCM_START?

File: crates/sound/src/vhu_sound.rs [hunk 10]
```diff
@@ -410,13 +410,58 @@ impl VhostUserSoundThread {
...
+
+                VIRTIO_SND_R_PCM_STOP => {
+                    let pcm_hdr = desc_chain
+                        .memory()
+                        .read_obj::<VirtioSoundPcmHeader>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+                    let stream_id = u32::from(pcm_hdr.stream_id);
+
+                    let state = audio_backend.get_stream_state(stream_id).unwrap();
+                    if state != VIRTIO_SND_R_PCM_PREPARE && state != VIRTIO_SND_R_PCM_START {
```
Comment (MatiasVara) [inline crates/sound/src/vhu_sound.rs:450]: if we never set the state to VIRTIO_SND_R_PCM_STOP then this check is not required.

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 11]
```diff
@@ -400,18 +403,46 @@ impl AudioBackend for PwBackend {
...
+    fn stop(&self, stream_id: u32) -> Result<()> {
+        debug!("pipewire stop");
+        self.thread_loop.lock();
+        let mut stream_states = self.stream_states.write().unwrap();
+        let stream_hash = self.stream_hash.read().unwrap();
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.set_active(false).expect("could not stop stream");
+        } else {
+            return Err(Error::StreamWithIdNotFound(stream_id));
+        };
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pw_backend.rs]:
```suggestion
        if let Some(stream) = stream_hash.get(&stream_id) {
            stream.set_active(false).expect("could not stop stream");
            stream_states[stream_id as usize] = VIRTIO_SND_R_PCM_STOP;
            self.thread_loop.unlock();
            Ok(())
        } else {
            Err(Error::StreamWithIdNotFound(stream_id))
        }
```
I would write it like this instead

File: crates/sound/src/vhu_sound.rs [hunk 12]
```diff
@@ -410,13 +410,57 @@ impl VhostUserSoundThread {
...
+                    let state = audio_backend.get_stream_state(stream_id)?;
+                    if state != VIRTIO_SND_R_PCM_PREPARE && state != VIRTIO_SND_R_PCM_START {
+                        error!(
+                            "Command {} is not allowed at {}",
+                            VIRTIO_SND_R_PCM_STOP, state
+                        );
+                        response = VirtioSoundHeader {
+                            code: VIRTIO_SND_S_NOT_SUPP.into(),
+                        };
+                    } else if audio_backend.stop(stream_id).is_err() {
```
Comment (bilelmoussaoui) [inline crates/sound/src/vhu_sound.rs]: You are losing the error information here which would make it more difficult to debug, i would instead write `if let Err(err) = audio_backend.stop(stream_id)` and log the error as well

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 13]
```diff
@@ -400,19 +403,47 @@ impl AudioBackend for PwBackend {
         if let Some(stream) = stream_hash.get(&stream_id) {
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pw_backend.rs:403]: I am somewhat not sure about this,  as calling the release function with a stream_id that already got disconnected will crash with `could not disconnect stream` :) instead you also have to make sure to remove the stream_id from stream/hash & stream_listener

File: crates/sound/src/audio_backends/pw_backend.rs [hunk 14]
```diff
@@ -400,19 +403,47 @@ impl AudioBackend for PwBackend {
...
         Ok(())
     }
+
+    fn start(&self, stream_id: u32) -> Result<()> {
+        debug!("pipewire start");
+        self.thread_loop.lock();
+        let mut stream_states = self.stream_states.write().unwrap();
+        let stream_hash = self.stream_hash.read().unwrap();
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.set_active(true).expect("could not start stream");
```
Comment (bilelmoussaoui) [inline crates/sound/src/audio_backends/pw_backend.rs:422]: Should this really panic or calling `stream.set_active(true)?;` is more appropriate ?

File: crates/sound/src/vhu_sound.rs [hunk 15]
```diff
@@ -410,13 +410,57 @@ impl VhostUserSoundThread {
...
+                    if state != VIRTIO_SND_R_PCM_PREPARE && state != VIRTIO_SND_R_PCM_START {
+                        error!(
+                            "Command {} is not allowed at {}",
+                            VIRTIO_SND_R_PCM_STOP, state
+                        );
+                        response = VirtioSoundHeader {
+                            code: VIRTIO_SND_S_NOT_SUPP.into(),
+                        };
+                    } else if let Err(err) = audio_backend.stop(stream_id) {
+                        error!("IO error during stop() {}", err);
```
Comment (bilelmoussaoui) [inline crates/sound/src/vhu_sound.rs:459]:
```suggestion
                        error!("IO error during stop() {err}");
```
I am sure clippy complains about this, same for others

File: crates/sound/src/audio_backends/pw_backend.rs (hunk 13 above)
Comment (bilelmoussaoui) [reply crates/sound/src/audio_backends/pw_backend.rs:403]: If you remove the item from the map, the std::mem::drop call would no longer be needed as it is automatically done

File: crates/sound/src/audio_backends/pw_backend.rs (hunk 14 above)
Comment (dorindabassey) [reply crates/sound/src/audio_backends/pw_backend.rs:422]: we want it to panic if it fails to set_active(true), i don't think `?` is the right thing to do in this case.

File: crates/sound/src/vhu_sound.rs (hunk 15 above)
Comment (dorindabassey) [reply crates/sound/src/vhu_sound.rs:459]: clippy doesn't complain about it.

Comment (bilelmoussaoui) [reply crates/sound/src/vhu_sound.rs:459]: You have an older rust version installed then.

Comment (dorindabassey) [reply crates/sound/src/vhu_sound.rs:459]: i just updated my rust and clippy doesn't complain about it

**Review by MatiasVara — APPROVED** [review]

File: crates/sound/src/vhu_sound.rs (hunk 15 above)
Comment (bilelmoussaoui) [reply crates/sound/src/vhu_sound.rs:459]: Interesting, maybe the lint is no longer enabled by default. Nvm

**Review by bilelmoussaoui — APPROVED** [review]

**Pull request merged**